framework/include/cppmicroservices/BundleInitialization.h

```diff
@@ -57,19 +61,26 @@ namespace cppmicroservices
  *    :cmake:command:`usFunctionGenerateBundleInit`.
  * \endrststar
  */
-#    define CPPMICROSERVICES_INITIALIZE_BUNDLE                                                                       \
-        std::atomic<cppmicroservices::BundleContextPrivate*> US_CTX_INS(US_BUNDLE_NAME) {};                          \
-                                                                                                                     \
-        extern "C" cppmicroservices::BundleContextPrivate* US_GET_CTX_FUNC(US_BUNDLE_NAME)();                        \
-        extern "C" cppmicroservices::BundleContextPrivate* US_GET_CTX_FUNC(US_BUNDLE_NAME)()                         \
-        {                                                                                                            \
-            return US_CTX_INS(US_BUNDLE_NAME).load();                                                                \
-        }                                                                                                            \
-                                                                                                                     \
-        extern "C" US_ABI_EXPORT void US_SET_CTX_FUNC(US_BUNDLE_NAME)(cppmicroservices::BundleContextPrivate * ctx); \
-        extern "C" US_ABI_EXPORT void US_SET_CTX_FUNC(US_BUNDLE_NAME)(cppmicroservices::BundleContextPrivate * ctx)  \
-        {                                                                                                            \
-            US_CTX_INS(US_BUNDLE_NAME).store(ctx);                                                                   \
+#    define CPPMICROSERVICES_INITIALIZE_BUNDLE                                                               \
+        struct {                                                                                             \
+            std::mutex mut;                                                                                  \
+            cppmicroservices::BundleContext* ctx = nullptr;                                                  \
+        } US_CTX_INS(US_BUNDLE_NAME);                                                                        \
```

Comment:
This is needed so that the two accesses on line 81 plus the access on line 82 all occur in one atomic transaction. Otherwise, for example, the `ctx` could be changed between the `if` and the `delete` on 81, which would result in deleting the wrong object.

Reply:
That does make sense. I recommend changing the design to remove the need for the mutex (and the need to call `delete`) entirely. `extern "C" US_ABI_EXPORT void US_SET_CTX_FUNC(US_BUNDLE_NAME)` is not called concurrently for the same bundle from within the CppMicroServices code base as far as I know.